File: 170. Two Sum III - Data structure design.py
class TwoSum:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.numbers = {}

    def add(self, number: int) -> None:
        """
        Add the number to an internal data structure..
        """
        if number in self.numbers:
            self.numbers[number] += 1
        else:
            self.numbers[number] = 1

    def find(self, value: int) -> bool:
        """
        Find if there exists any pair of numbers which sum is equal to the value.
        """
        for n in self.numbers:
            if value - n in self.numbers and (value != 2*n or self.numbers[n] >= 2):
                return True
        return False


# Your TwoSum object will be instantiated and called as such:
# obj = TwoSum()
# obj.add(number)
# param_2 = obj.find(value)

File: 170. Two Sum III - Data structure design.java
/**
 * Use hashmap to record numbers, 
 * and check if value-i in the map or if it is the same check for if we have duplicates
 */
class TwoSum {
    HashMap<Integer, Integer> numCount;

    /** Initialize your data structure here. */
    public TwoSum() {
        numCount = new HashMap<>();
    }
    
    /** Add the number to an internal data structure.. */
    public void add(int number) {
        Integer i = numCount.get(number);
        if (i==null) {
            numCount.put(number, 1);
        } else {
            numCount.put(number, i+1);
        }
    }
    
    /** Find if there exists any pair of numbers which sum is equal to the value. */
    public boolean find(int value) {
        for (int i : numCount.keySet()) {
            if (numCount.containsKey(value-i) && (value != 2*i || numCount.get(i) >= 2)) {
                return true;
            }
        }
        return false;
    }
}

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum obj = new TwoSum();
 * obj.add(number);
 * boolean param_2 = obj.find(value);
 */